core/src/main/java/com/netflix/msl/msg/MessageBuilder.java

```diff
@@ -1123,37 +815,37 @@ public MessageBuilder deletePeerServiceToken(final String name) {
     public Set<ServiceToken> getPeerServiceTokens() {
         return Collections.unmodifiableSet(new HashSet<ServiceToken>(peerServiceTokens.values()));
     }
-    
+
     /** MSL context. */
-    private final MslContext ctx;
-    
+    protected MslContext ctx;
```

Comment:
Keep `ctx` final and require it to be passed in from the protected constructor.

The current code is written under the assumption that `ctx` cannot be null, however since the code has been changed to sometimes set `ctx` from `initializeMessageBuilder()` with an expectation that it is called from a subclass or one of `MessageBuilder`'s public constructors, this assumption is no longer valid. However the code in the functions inside `MessageBuilder` is incompatible with that as currently written.

Adding checks for null `ctx` values has the problem of requiring you to add error handling (i.e. exceptions) to those functions since execution cannot succeed anymore, and all the yuckiness that would result from that new exception. Plus it moves detection of invalid object state from constructor into all the other random class methods (i.e. you are permitted to create an invalid object).

You avoid all of those problems by keeping `ctx` final and forcing it to be set all constructors.